# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/SharedArrayObject.cpp
# Commit: 85d47a33200c
# Full Hash: 85d47a33200c36a5f91f011f5713c34da79329ee
# Author: Julien Pages <jpages@mozilla.com>
# Date: 2025-09-15 21:44:14
# Regressor Bug: 1969509
# File Overlap Count: 7
# Overlapping Files: js/src/vm/ArrayBufferObject.cpp, js/src/wasm/WasmMemory.h, js/src/jit-test/tests/wasm/wasm-resizablearraybuffer-shared.js, js/src/jit-test/tests/wasm/memory64/basic.js, js/src/jit-test/tests/wasm/wasm-resizablearraybuffer.js
# Description:
#   Bug 1969509 - wasm: Update .maxByteLength to report the source wasm max pages in bytes. r=rhunt
#   
#   Following a spec change, 64-bit memory limits in wasm have been increased.
#   It's now possible for a wasm memory to have a limit bigger than the maximum
#   value of a 32-bits integer. To avoid an overflow, the maxByteLength getters
# ==============================================================================

diff -r 5860a981c750 -r 85d47a33200c js/src/vm/SharedArrayObject.cpp
--- a/js/src/vm/SharedArrayObject.cpp	Mon Sep 15 13:51:31 2025 +0000
+++ b/js/src/vm/SharedArrayObject.cpp	Mon Sep 15 14:01:27 2025 +0000
@@ -331,6 +331,28 @@
   MOZ_ASSERT(IsSharedArrayBuffer(args.thisv()));
   auto* buffer = &args.thisv().toObject().as<SharedArrayBufferObject>();
 
+  // Special case for wasm with potentially 64-bits memory.
+  // Manually compute maxByteLength to avoid an overflow on 32-bit machines.
+  if (buffer->isWasm()) {
+    auto* wasmBuffer = buffer->rawWasmBufferObject();
+
+    args.rval().setNumber(
+        double(wasmBuffer->wasmSourceMaxPages().value() * wasm::PageSize));
+    return true;
+  }
+  // Special case for wasm with potentially 64-bits memory.
+  // Manually compute the maxByteLength to avoid an overflow on 32-bit machines.
+  if (buffer->isWasm()) {
+    Pages sourceMaxPages = buffer->rawWasmBufferObject()->wasmSourceMaxPages();
+    uint64_t sourceMaxBytes = sourceMaxPages.byteLength64();
+
+    MOZ_ASSERT(sourceMaxBytes <=
+               wasm::PageSize * wasm::MaxMemory64PagesValidation);
+    args.rval().setNumber(double(sourceMaxBytes));
+
+    return true;
+  }
+
   // Steps 4-6.
   args.rval().setNumber(buffer->byteLengthOrMaxByteLength());
   return true;